src/shared/edit-util.c

```diff
@@ -0,0 +1,378 @@
+                const char *marker_start,
+                const char *marker_end,
+                char **ret_temp_filename,
+                unsigned *ret_edit_line) {
+
+        _cleanup_free_ char *temp = NULL;
+        unsigned line = 1;
+        int r;
+
+        assert(target_path);
+        assert(!comment_paths || (comment_paths && (marker_start || marker_end)));
+        assert(ret_temp_filename);
+
+        r = tempfn_random(target_path, NULL, &temp);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine temporary filename for \"%s\": %m", target_path);
+
+        r = mkdir_parents_label(target_path, 0755);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create parent directories for \"%s\": %m", target_path);
+
+        if (original_path) {
+                r = mac_selinux_create_file_prepare(target_path, S_IFREG);
+                if (r < 0)
+                        return r;
+
+                r = copy_file(original_path, temp, 0, 0644, 0, 0, COPY_REFLINK);
+                if (r == -ENOENT) {
+                        r = touch(temp);
+                        mac_selinux_create_file_clear();
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create temporary file \"%s\": %m", temp);
+                } else {
+                        mac_selinux_create_file_clear();
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create temporary file for \"%s\": %m", target_path);
+                }
+        }
+
+        if (comment_paths) {
+                _cleanup_free_ char *target_contents = NULL;
+                _cleanup_fclose_ FILE *f = NULL;
+
+                r = mac_selinux_create_file_prepare(target_path, S_IFREG);
+                if (r < 0)
+                        return r;
+
+                f = fopen(temp, "we");
+                mac_selinux_create_file_clear();
+                if (!f)
+                        return log_error_errno(errno, "Failed to open temporary file \"%s\": %m", temp);
+
+                if (fchmod(fileno(f), 0644) < 0)
+                        return log_error_errno(errno, "Failed to change mode of temporary file \"%s\": %m", temp);
+
+                r = read_full_file(target_path, &target_contents, NULL);
+                if (r < 0 && r != -ENOENT)
+                        return log_error_errno(r, "Failed to read target file \"%s\": %m", target_path);
+
+                fprintf(f,
+                        "### Editing %s\n"
+                        "%s\n"
+                        "\n"
+                        "%s%s"
+                        "\n"
+                        "%s\n",
+                        target_path,
+                        marker_start,
+                        strempty(target_contents),
+                        target_contents && endswith(target_contents, "\n") ? "" : "\n",
+                        marker_end);
```

Comment:
In the assertion at the beginning of the function, `marker_start` _or_ `marker_end` is requested. Hence, one of them can be NULL. Is it intended? If so, please do not print `(null)`.